-Wsevere – erroring warnings (#571)
ghc-proposals/ghc-proposals: proposals/0000-severe-warnings.rst

-Wsevere – erroring warnings
============================

.. author:: Joachim Breitner
.. date-accepted::
.. ticket-url::
.. implemented::
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/571>`_.
.. contents::

> It seems helpful if some warning categories (e.g. ``missing-methods``) to be
> *errors* by default (as if ``-Werror=missing-methods`` was given), but still
> allow them to be turns to warnings (``-Wwarn=missing=methods``) or quiet
> (``-Wno-missing-methods``). This proposal
> adds a new warning group ``-Wsevere`` (like ``-Wextra`` or ``-Wdefault``) for such warnings
> and upgrades ``missing-methods``,  ``missing-fields`` from ``-Wdefault`` to ``-Wsevere``
> (fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_)

Reviewer: ```suggestion
It seems helpful for some warnings (e.g. ``missing-methods``) to be
*errors* by default (as if ``-Werror=missing-methods`` was given), but still
allow them to be turned into warnings (``-Wwarn=missing=methods``) or being disabled
(``-Wno-missing-methods``). This proposal
adds a new warning group ``-Wsevere`` (like ``-Wextra`` or ``-Wdefault``) for such warnings
and upgrades ``missing-methods``,  ``missing-fields`` from ``-Wdefault`` to ``-Wsevere``
(fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_).
```

Motivation
----------

* We can start for the motivation for making ``missing-methods`` an error by default, copying from @phadej’s
  `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_:
  
    Currently omitting method implementations in type-class instances
    is just a warning, but it is enabled by default.
    Therefore one could argue that adding a new method to existing
    type-class is not a breaking change.
    However, it very much is, but

    *  ecosystem is not forced to catch up in any timely matter
    *  the breakage is hard to assess

  and (to me even more compelling):
  
    FWIW, base-4.18.0.0 is going to have

    Add default implementation of `(<>)` in terms of `sconcat` and `mempty` in terms of `mconcat`.
    which makes it possible to write::

      instance Semigroup Foo
      instance Monoid Foo

    and that instance's mempty will loop.
    Currently GHC will just warn if someone writes such (new) code by accident.

>   It seems to me we are doing our users a service if we make that particular foot a little less easy to fire accidentially

Reviewer: ```suggestion
  It seems to me we are doing our users a service if we make that particular foot a little less easy to fire in accidentially.
```

Author: ```suggestion
  It seems to me we are doing our users a service if we make that particular foot gun
  a little less easy to fire accidentially.
```

> * The motivation for ``-Werror=missing-fields`` follows by dualty – instances are just records for implicitly passed parameters, in a way.

Reviewer: ```suggestion
* The motivation for ``-Werror=missing-fields`` follows by duality – instances are just records for implicitly passed parameters, in a way.
```

* From the need to have *some* warnings being errors by default follows the need for the suitable infrastructure, hence the
  ``-Wsevere`` warning group.
  

> * Given that we have custom warning categories (`x-foo`, `#454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_),

Reviewer 2: ```suggestion
* Given that we have custom warning categories (``x-foo``, `#541 <https://github.com/ghc-proposals/ghc-proposals/pull/541>`_),
```

  we anticipate that users of custom warning categories will want a way to add “their” warning group to ``-Wsevere`` as well.
  
  Because there is no registry of custom categories, we let the prefix indicate the group. The prefix ``x-`` indicates a custom warning in ``-Wdefault``.

>   This proposal allows the prefix ``xs-` to indicate that this group should be part of the severe group.

Reviewer 2: ```suggestion
  This proposal allows the prefix ``xs-`` to indicate that this group should be part of the severe group.
```

   
  
Proposed Change Specification
-----------------------------

> * GHC learns a new warning group flag ``-Wsevere``, next to the existing ``-Wdefault``, ``-Wextra``, `-Wall`` and ``-Wcompat``.
>   See _“5.2. Warnings and sanity-checking” <https://downloads.haskell.org/ghc/latest/docs/users_guide/using-warnings.html>_.

Reviewer 2: ```suggestion
* GHC learns a new warning group flag ``-Wsevere``, next to the existing ``-Wdefault``, ``-Wextra``, ``-Wall`` and ``-Wcompat``.
  See `“5.2. Warnings and sanity-checking” <https://downloads.haskell.org/ghc/latest/docs/users_guide/using-warnings.html>`_.
```

  
  It behaves like the other groups in that ``-Wno=severe``, ``-Wwarn=severe`` and ``-Werror=severe`` behave as if the flag

>   had been specified for each warning in the groups.

Reviewer 2: This should be `-Wno-severe` with a `-` rather than `=`, for consistency (and similarly below). And actually most of the groups *don't* support `-Wwarn=...` or `-Werror=...` in current releases, but I consider this a bug and am fixing it (https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9679).

Author: I keep making assumptions about how things are only to learn that it's only how they should be :-)

Is there a reason we have `=` after `error`, but not after `warn` and `no`?

Reviewer 2: Actually it is `warn=` and `error=` (and secretly also `no-error=`) but the odd one out is `no-`. I don't know how that design was settled on. :grin:

Author: Ok, whatever :-)

  
* The warning group is initialized with the following warning categories:

  * ``missing-methods``
  * ``missing-fields``

*  The set of allowed names of custom warning categories (`#454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_) is

>    extended by those prefixed with ``xs-*` (for “custom extended – severe”). The flags

Reviewer 2: I like this approach, but I'm wondering how the `-Wextended-warnings` group introduced in #541 relates to the new `xs-` prefix. Perhaps `-Wextended-warnings` includes all categories regardless of prefix, but we add a new `-Wsevere-extended-warnings` group that includes only the `xs-` prefixed categories? Thus the default state can easily be described as having `-Wextended-warnings` in `-Wdefault` and `-Wsevere-extended-warnings` in `-Wsevere`.

Do we anticipate wanting to extend this to other warning groups, e.g. `-Wall`?

Author: I am not sure if we even need the all-custom-warnings-flag. If we do, adding another for all severe ones is reasonable, I think.

I am not sure if there is demand for off-by-default custom warnings, but I'd there is, `xa-` or something like that would work, would it?

>    extended by those prefixed with ``xs-*` (for “custom extended – severe”). The flags

Reviewer: ```suggestion
   extended by those prefixed with ``xs-*`` (for “custom extended – severe”). The flags
```

> *  The set of allowed names of custom warning categories (`#454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_) is
>    extended by those prefixed with ``xs-*` (for “custom extended – severe”). The flags
>    ``-Wno=severe``, ``-Wwarn=severe`` and ``-Werror=severe``
>    also affect all custom warnigns with ah name starting in ``xs-*``.

Reviewer 2: ```suggestion
*  The set of allowed names of custom warning categories (`#541 <https://github.com/ghc-proposals/ghc-proposals/pull/541>`_) is
   extended by those prefixed with ``xs-*` (for “custom extended – severe”). The flags
   ``-Wno-severe``, ``-Wwarn=severe`` and ``-Werror=severe``
   also affect all custom warnings with a name starting in ``xs-*``.
```

 

Examples
--------
TODO

Effect and Interactions
-----------------------
TODO

Costs and Drawbacks
-------------------
In terms of GHC development, this is a modest extension of the existing warning category and group infrastructure.

In terms of breakage, @phadej has made an impact analysis at <https://github.com/ghc-proposals/ghc-proposals/issues/544#issue-1410125536>.

Alternatives
------------

* Naming the group.

  I suggest to use the name ``severe``, which does not have ``error`` in it (e.g. ``errors-by-default``), because else
  ``-Werror=errors-by-default`` or something looks kinda strange, and after someone says ``-Wwarn=severe``, the ``severe`` group still
  exists, but it not an error.
  

>   I briefly considered ``-Wfatal`, but that’s a lie – these errors are *not* ``fatal``, else we couldn’t turn them

Reviewer 2: ```suggestion
  I briefly considered ``-Wfatal``, but that’s a lie – these errors are *not* ``fatal``, else we couldn’t turn them
```

  into warnings.
  
* We could add more warnings to the group right away.

* In particular, we could make ``partial-type-signatures`` a normal warning, in ``-Wsevere`` by default, and ``-XPartialTypeSignatures``
  becomes a synonym for ``-Wwarn=partial-type-signatures``.
  

>   (Or should ``-XPartialTypeSignatures`` be a flag that changes the *warning group* of ``partial-type-signatures``? Not sure).

Reviewer 2: I think some form of this is probably a good idea, but I suggest to keep the current proposal simple and deal with this in a follow up proposal at some point, once the basic `-Wsevere` group is established.

(``-Wno-missing-methods``). This proposal
adds a new warning group ``-Wsevere`` (like ``-Wextra`` or ``-Wdefault``) for such warnings
and upgrades ``missing-methods``,  ``missing-fields`` from ``-Wdefault`` to ``-Wsevere``
(fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_).

Motivation
----------

* We can start for the motivation for making ``missing-methods`` an error by default, copying from @phadej’s
  `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_:
  
    Currently omitting method implementations in type-class instances
    is just a warning, but it is enabled by default.
    Therefore one could argue that adding a new method to existing
    type-class is not a breaking change.
    However, it very much is, but

    *  ecosystem is not forced to catch up in any timely matter
    *  the breakage is hard to assess

  and (to me even more compelling):
  
    FWIW, base-4.18.0.0 is going to have

    Add default implementation of `(<>)` in terms of `sconcat` and `mempty` in terms of `mconcat`.
    which makes it possible to write::

      instance Semigroup Foo
      instance Monoid Foo

    and that instance's mempty will loop.

>     Currently GHC will just warn if someone writes such (new) code by accident.

Reviewer 3: I'm confused by this, if the methods have default implementations, will GHC really warn?

Author: Good point. Maybe a MINIMAL pragma is involved (but they don't span multiple classes, do they). @phadej?

Reviewer 4: with MINIMAL pragma GHC warns. That's the point of MINIMAL pragma, to allow circular default definitions but make compiler still warn.

Reviewer 3: Ah, right, so the circular definitions don't actually span multiple classes. Makes sense!

  becomes a synonym for ``-Wwarn=partial-type-signatures``.
  
  (Or should ``-XPartialTypeSignatures`` be a flag that changes the *warning group* of ``partial-type-signatures``? Not sure).

* An additional motivation is the currently circulating idea that some features that are language *extensions* right now (``-XFoo``) can
  simply become part of the default “normal” language, together with a new warning category (``foo``) that’s off by default normally, and 
  the extension flags ``-XNoFoo`` or ``-XHaskell2010`` simply have the same effect as ``-Werror=foo``.
  
  Although now that I write it out, it seems that we don't actually need ``-Wsevere`` for that. Maybe it’s useful to *strongly deprecate* features,
  by introducing a warning about their use, and raising its severity from ``-Wcompat`` to ``-Wdefault`` to ``-Werror`` over time.

* We could guard this change behind a suitable language extension, so that ``Haskell2010`` stays untouched. It could be the default eventually, but
  would not affect code under ``Haskell2010`` or ``GHC2021``.
  

>   It would set precedent for langauge extensions changing the default mode (warning vs. errors) of warnings, and I’d propose that the semantics

Reviewer 3: ```suggestion
  It would set precedent for language extensions changing the default mode (warning vs. errors) of warnings, and I’d propose that the semantics
```

  would be that all language flags (``-X``) are processed, from that the default on/off and error/warn sets are derived, and then all ``-W`` flags
  are processed, so that ``-X`` and ``-W`` flags commute.
  

>   But as explained in the motivation the motiviation comes from imposing this more rigid discipline on existing code, and so no language extension

Reviewer: ```suggestion
  But as explained in the motivation the motivation comes from imposing this more rigid discipline on existing code, and so no language extension
```

(fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_).

Motivation
----------

* We can start for the motivation for making ``missing-methods`` an error by default, copying from @phadej’s
  `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_:
  
    Currently omitting method implementations in type-class instances
    is just a warning, but it is enabled by default.
    Therefore one could argue that adding a new method to existing
    type-class is not a breaking change.
    However, it very much is, but

    *  ecosystem is not forced to catch up in any timely matter
    *  the breakage is hard to assess

  and (to me even more compelling):
  
    FWIW, base-4.18.0.0 is going to have

    Add default implementation of `(<>)` in terms of `sconcat` and `mempty` in terms of `mconcat`.
    which makes it possible to write::

      instance Semigroup Foo
      instance Monoid Foo

    and that instance's mempty will loop.
    Currently GHC will just warn if someone writes such (new) code by accident.

  It seems to me we are doing our users a service if we make that particular foot gun
  a little less easy to fire accidentally.
  

> * The motivation for ``-Werror=missing-fields`` follows by duality – instances are just records for implicitly passed parameters, in a way.

Reviewer 2: The motivation for making `missing-fields` severe feels slightly weaker than `missing-methods`, because failure to supply a field leads to a reasonably clear runtime exception, whereas failure to supply a method can lead to an infinite loop. I don't think it's unreasonable, but we could imagine doing one and not the other.

  3 |   foo = Foo { a = 1 }
    |         ^^^^^^^^^^^^^
  $ echo $?
  1
  $ ghc -Wwarn=missing-fields Test.hs
  [1 of 1] Compiling Test             ( Test.hs, Test.o )
  
  Test.hs:3:9: warning: [-Wmissing-fields]
      • Fields of ‘Foo’ not initialised:
          b :: Bool
      • In the expression: Foo {a = 1}
        In an equation for ‘foo’: foo = Foo {a = 1}
    |
  3 |   foo = Foo { a = 1 }
    |         ^^^^^^^^^^^^^
  $ echo $?
  0
  $ ghc -Wno-missing-fields Test.hs
  $ echo $?
  0

Disabling the ``default`` warnings, but not the ``severe`` warnings, will leave the ``severe`` warnings on::

  $ ghc Test.hs -Wno-default
  [1 of 1] Compiling Test             ( Test.hs, Test.o )
  
  Test.hs:3:9: error: [-Wmissing-fields, -Werror=missing-fields]
      • Fields of ‘Foo’ not initialised:
          b :: Bool
      • In the expression: Foo {a = 1}
        In an equation for ‘foo’: foo = Foo {a = 1}
    |
  3 |   foo = Foo { a = 1 }
    |         ^^^^^^^^^^^^^
  

Effect and Interactions
-----------------------
None yet.

Costs and Drawbacks
-------------------
In terms of GHC development, this is a modest extension of the existing warning category and group infrastructure.

> In terms of breakage, @phadej has made an impact analysis at <https://github.com/ghc-proposals/ghc-proposals/issues/544#issue-1410125536>.

Reviewer 2: This comment talks about `missing-methods`. It wasn't immediately obvious to me that he also kindly provided an analysis for `missing-fields` (https://github.com/ghc-proposals/ghc-proposals/issues/544#issuecomment-1279948737), so perhaps link to that here?

Given the likely impact I think the proposal text could do with backporting the "Backwards Compatibility" section from the latest version of the template, explaining that (a) there will be breakage and (b) the breakage is justified.

Author: I added the section with, very little text. If you want, feel free to elaborate (just commit directly, or use the suggested edit feature :-))

Reviewer 2: Looks good. :-) Ready to submit?
